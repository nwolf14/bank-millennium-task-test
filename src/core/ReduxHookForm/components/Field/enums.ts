export enum InputValueType {
  boolean,
  number,
  string,
}
