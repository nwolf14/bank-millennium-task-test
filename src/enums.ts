export enum KeyboardKeys {
  backspace = 'Backspace',
  delete = 'Delete',
  enter = 'Enter',
  escape = 'Escape',
  shift = 'Shift',
}

export enum MousePosition {
  northWest = 'northWest',
  northEast = 'northEast',
  southWest = 'southWest',
  southEast = 'southEast',
}
