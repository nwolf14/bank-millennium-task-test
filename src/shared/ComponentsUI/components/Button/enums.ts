export enum Color {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum Size {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum Variant {
  text = 'text',
  contained = 'contained',
  outlined = 'outlined',
}
