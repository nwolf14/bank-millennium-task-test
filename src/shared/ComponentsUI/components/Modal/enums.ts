export enum ModalId {
  storyBookModal = 'stroryBookModal',
}
