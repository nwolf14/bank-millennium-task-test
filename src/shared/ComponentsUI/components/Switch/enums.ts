export enum Size {
  small = 'small',
  medium = 'medium',
}
