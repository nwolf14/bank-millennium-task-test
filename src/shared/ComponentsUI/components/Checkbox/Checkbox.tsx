import camelCase from 'lodash/camelCase';
import classnames from 'classnames';
import isArray from 'lodash/isArray';
import {
  ChangeEvent,
  forwardRef,
  MutableRefObject,
  useRef,
  useState,
} from 'react';

// components
import CheckboxIconWrapper, {
  TProps as TCheckboxIconWrapperProps,
} from './components/CheckboxIconWrapper/CheckboxIconWrapper';
import CheckboxInput, {
  TProps as TCheckboxInputProps,
} from './components/CheckboxInput/CheckboxInput';
import CheckboxLabel, {
  TProps as TCheckboxLabelProps,
} from './components/CheckboxLabel/CheckboxLabel';
import ValidationMessage from '../../../ValidationMessage/ValidationMessage';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameCheckbox, classNames } from './classNames';
import { INPUT_STATES_INITIAL } from '../../constants';
import { Margin } from '../../enums';
import { Size } from './enums';
import { THookType as TMetaProps } from '../../../../core/ReduxHookForm/components/Field/hooks/useMetaProps';

// utils
import { getErrorMessage } from '../../../../utils/form/inputStates';
import { getRandomKey } from '../../../../utils/getRandomKey';

// styles
import './checkbox.scss';

export type TProps = Pick<
  TCheckboxIconWrapperProps,
  'CheckedIcon' | 'classNameIconWrapper' | 'UncheckedIcon'
> &
  Pick<TCheckboxInputProps, 'classNameInput'> &
  Pick<TCheckboxLabelProps, 'classNameLabel'> & {
    className?: string;
    defaultValue?: boolean;
    disabled?: boolean;
    disabledPulseEffect?: boolean;
    errorMessage?: string;
    forcedFocus?: boolean;
    forcedHover?: boolean;
    index?: number;
    label?: string;
    margin?: Margin;
    meta?: TMetaProps;
    onChange?: (checked: boolean) => void;
    setValueGroup?: (value: boolean, index: number) => void;
    size?: Size;
    style?: { [key: string]: number | string };
    value?: boolean;
  };

export const Checkbox = forwardRef<HTMLInputElement, TProps>(
  (
    {
      CheckedIcon,
      className = '',
      classNameIconWrapper,
      classNameInput,
      classNameLabel,
      defaultValue = false,
      disabled = false,
      disabledPulseEffect = false,
      errorMessage = '',
      forcedFocus = false,
      forcedHover = false,
      index = 0,
      label = '',
      margin = Margin.marginBottom,
      meta = {},
      onChange = null,
      setValueGroup = null,
      size = Size.medium,
      style = {},
      UncheckedIcon,
      value: controlledValue = null,
    },
    ref
  ) => {
    const { classNamesWithTheme } = useTheme(classNames);
    const [inputStates, setInputStates] = useState(INPUT_STATES_INITIAL);
    const [pulseElements, setPulseElements] = useState<Array<string>>([]);
    const [value, setValue] = useState<boolean>(defaultValue);
    const inputRef = useRef<HTMLInputElement>(
      null
    ) as MutableRefObject<HTMLInputElement>;
    const error = getErrorMessage(meta as TMetaProps) || errorMessage;

    const onChangeHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const {
        target: { checked: value },
      } = event;

      if (!disabledPulseEffect) {
        setPulseElements([...pulseElements, getRandomKey(pulseElements)]);
      }

      if (onChange) {
        onChange(value);
      }

      if (setValueGroup) {
        setValueGroup(value, index);
      }

      setValue(value);
    };

    const assignRef = (input: HTMLInputElement): void => {
      if (ref) {
        const definedRef = ref as MutableRefObject<HTMLInputElement>;

        if (isArray(definedRef.current)) {
          definedRef.current[index] = input;
        } else {
          definedRef.current = input;
        }
      } else {
        inputRef.current = input;
      }
    };

    const getValue = (): boolean =>
      controlledValue !== null ? controlledValue : value;

    return (
      <div
        className={classnames(
          classNamesWithTheme[classNameCheckbox].name,
          className,
          classNamesWithTheme[classNameCheckbox].modificators[camelCase(margin)]
        )}
        style={style}
      >
        <div className={classNamesWithTheme.wrapper}>
          <CheckboxInput
            value={getValue()}
            classNameInput={classNameInput}
            disabled={disabled}
            inputStates={inputStates}
            onChange={onChangeHandler}
            ref={(input) => assignRef(input as HTMLInputElement)}
            setInputStates={setInputStates}
            size={size}
          />
          <CheckboxIconWrapper
            value={getValue()}
            CheckedIcon={CheckedIcon}
            classNameIconWrapper={classNameIconWrapper}
            disabled={disabled}
            error={error}
            forcedFocus={forcedFocus}
            forcedHover={forcedHover}
            inputStates={inputStates}
            pulseElements={pulseElements}
            setPulseElements={setPulseElements}
            size={size}
            UncheckedIcon={UncheckedIcon}
          />
          <CheckboxLabel
            classNameLabel={classNameLabel}
            disabled={disabled}
            index={index}
            label={label}
            ref={ref || inputRef}
          />
        </div>
        <ValidationMessage error={error} />
      </div>
    );
  }
);

export default Checkbox;
