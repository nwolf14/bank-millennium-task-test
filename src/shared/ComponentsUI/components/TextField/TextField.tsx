import camelCase from 'lodash/camelCase';
import classnames from 'classnames';
import { ChangeEvent, forwardRef, useEffect, useState } from 'react';

// components
import ValidationMessage from '../../../ValidationMessage/ValidationMessage';
import TextFieldInputWrapper, {
  TProps as TTextFieldInputWrapperProps,
} from './components/TextFieldInputWrapper/TextFieldInputWrapper';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameTextField, classNames } from './classNames';
import { HEIGHTS } from './constants';
import { Margin } from '../../enums';
import { Size } from './enums';
import { INPUT_STATES_INITIAL } from '../../constants';
import { THookType as TMetaProps } from '../../../../core/ReduxHookForm/components/Field/hooks/useMetaProps';
import { TObject } from '../../../../types';

// styles
import './text-field.scss';

// utils
import { getErrorMessage } from '../../../../utils/form/inputStates';

export type TProps = Omit<
  TTextFieldInputWrapperProps,
  | 'error'
  | 'height'
  | 'inputStates'
  | 'multiline'
  | 'onBlur'
  | 'onChange'
  | 'onFocus'
  | 'required'
  | 'rows'
  | 'setInputStates'
  | 'size'
  | 'value'
> & {
  className?: string;
  errorMessage?: string;
  defaultValue?: boolean | number | string | Array<string>;
  fullWidth?: boolean;
  helperText?: string;
  margin?: Margin;
  maxRows?: number;
  meta?: TMetaProps;
  minRows?: number;
  multiline?: boolean;
  onBlur?: (value: number | string) => void;
  onChange?: (value: number | string) => void;
  onFocus?: (value: number | string) => void;
  required?: boolean;
  size?: Size;
  style?: TObject<number | string>;
  value?: number | string;
};

export const TextField = forwardRef<HTMLInputElement, TProps>(
  (
    {
      className = '',
      defaultValue = '',
      errorMessage = '',
      fullWidth = false,
      helperText,
      margin = Margin.marginBottom,
      meta = {},
      multiline = false,
      required = false,
      maxRows = 0,
      minRows = 0,
      onBlur = null,
      onChange = null,
      onFocus = null,
      size = Size.medium,
      style = {},
      value: controlledValue = null,
      ...restProps
    },
    ref
  ) => {
    const [height, setHeight] = useState(`${HEIGHTS[size] * (minRows || 1)}px`);
    const [inputStates, setInputStates] = useState(INPUT_STATES_INITIAL);
    const [rows, setRows] = useState(minRows || 1);
    const [value, setValue] = useState(defaultValue);
    const { classNamesWithTheme } = useTheme(classNames);
    const error = getErrorMessage(meta as TMetaProps) || errorMessage;

    const setInputHeight = (value: string): void => {
      const amountLines = value.split(/\r?\n|\r/).length;

      if (!minRows && !maxRows) {
        setRows(amountLines);
        setHeight(`${HEIGHTS[size] * amountLines}px`);
      } else if (amountLines <= maxRows) {
        if (amountLines <= minRows) {
          setHeight(`${HEIGHTS[size] * minRows}px`);
        } else {
          setHeight(`${HEIGHTS[size] * amountLines}px`);
        }
      }
    };

    const onBlurHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setInputStates({ ...inputStates, focus: false });

      if (onBlur) {
        onBlur(value);
      }
    };

    const onChangeHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setValue(value);

      if (multiline) {
        setInputHeight(value);
      }

      if (onChange) {
        onChange(value);
      }
    };

    const onFocusHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setInputStates({ ...inputStates, focus: true });

      if (onFocus) {
        onFocus(value);
      }
    };

    const getValue = (): string | number | boolean | Array<string> =>
      controlledValue !== null ? controlledValue : value;

    useEffect(() => {
      if (getValue() && multiline) {
        setInputHeight(defaultValue as string);
      }
    }, []);

    return (
      <div
        className={classnames(
          classNamesWithTheme[classNameTextField].name,
          {
            [classNamesWithTheme[classNameTextField].modificators.fullWidth]:
              fullWidth,
          },
          classNamesWithTheme[classNameTextField].modificators[
            camelCase(margin)
          ],
          className
        )}
        style={style}
      >
        <TextFieldInputWrapper
          error={error}
          height={height}
          inputStates={inputStates}
          multiline={multiline}
          onBlur={onBlurHandler}
          onChange={onChangeHandler}
          onFocus={onFocusHandler}
          ref={ref}
          required={required}
          rows={rows}
          setInputStates={setInputStates}
          size={size}
          value={getValue() as string | number}
          {...restProps}
        />
        <ValidationMessage error={error} helperText={helperText} />
      </div>
    );
  }
);

export default TextField;
