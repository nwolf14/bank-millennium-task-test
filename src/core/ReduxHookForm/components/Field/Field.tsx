import { defer } from 'lodash';
import {
  FC,
  forwardRef,
  ForwardRefExoticComponent,
  ReactNode,
  RefObject,
  useEffect,
} from 'react';
import { useDispatch, useSelector } from 'react-redux';

// hooks
import { useInputProps } from './hooks/useInputProps';
import {
  THookType as TFieldMetaProps,
  useMetaProps,
} from './hooks/useMetaProps';
import { useValidators } from './hooks/useValidators';

// others
import { TAsyncValidator, TSyncValidator } from '../../types';
import { TProps as TTextFieldProps } from '../../../../shared/ComponentsUI/components/TextField/TextField';
import { TProps as TCheckboxProps } from '../../../../shared/ComponentsUI/components/Checkbox/Checkbox';
import { TProps as TRadioButtonProps } from '../../../../shared/ComponentsUI/components/RadioButton/RadioButton';
import { TProps as TSwitchProps } from '../../../../shared/ComponentsUI/components/Switch/Switch';

// store
import { getFieldSelectorCreator } from '../../../../store/reduxHookForm/selectors';
import { initField } from '../../../../store/reduxHookForm/actions';
import { InputValueType } from './enums';

export type TProps = Partial<TCheckboxProps> &
  Partial<TRadioButtonProps> &
  Partial<TSwitchProps> &
  Partial<TTextFieldProps> & {
    afterSubmit?: () => void;
    asyncValidators?: Array<TAsyncValidator>;
    beforeSubmit?: () => void;
    children?: (
      metaProps: Partial<TFieldMetaProps>,
      restProps?: any
    ) => ReactNode;
    Component?: FC<any> | ForwardRefExoticComponent<any>;
    data?: { [key: string]: any };
    formatOnBlur?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnChange?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnFocus?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnInit?: (
      value: boolean | number | string | Array<string>
    ) => boolean | number | string;
    formName?: string;
    inputValueType?: InputValueType;
    name: string;
    parse?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    ref?: RefObject<HTMLInputElement>;
    subscriptionFields?: Array<string>;
    syncValidators?: Array<TSyncValidator>;
    touched?: boolean;
    touchedAfterClick?: boolean;
    visited?: boolean;
  };

const Field = forwardRef<HTMLInputElement, TProps>(
  (
    {
      afterSubmit,
      asyncValidators = [],
      beforeSubmit,
      children,
      Component,
      data,
      defaultValue,
      formatOnBlur,
      formatOnChange,
      formatOnFocus,
      formatOnInit,
      formName = '',
      inputValueType = InputValueType.string,
      name = '',
      parse,
      subscriptionFields = [],
      syncValidators = [],
      touched: initialTouched = false,
      touchedAfterClick = false,
      visited: initialVisited = false,
      ...restProps
    },
    ref
  ) => {
    const field = useSelector(getFieldSelectorCreator(formName, name));
    const { touched, value = '', visited } = field || {};
    const dispatch = useDispatch();
    const inputProps = useInputProps(formName, name);
    const metaProps = useMetaProps(formName, name);
    const { getSyncErrors, updateAsyncValidators, updateSyncValidators } =
      useValidators(
        asyncValidators,
        formName,
        name,
        subscriptionFields,
        syncValidators
      );
    const wrapperRestProps = {
      ...inputProps,
      ...restProps,
      ...(ref ? { ref: ref } : {}),
    };

    const getInitialValue = (): boolean | number | string | Array<string> => {
      const initValue = formatOnInit
        ? formatOnInit(defaultValue || '')
        : defaultValue;

      switch (inputValueType) {
        case InputValueType.boolean:
          return initValue || false;
        case InputValueType.number:
          return initValue || 0;
        default:
          return initValue || '';
      }
    };

    useEffect(() => {
      if (formName) {
        const initialValue = getInitialValue();

        defer(() => {
          dispatch(
            initField({
              field: {
                afterSubmit,
                active: false,
                asyncErrors: [],
                beforeSubmit,
                data,
                formatOnBlur,
                formatOnChange,
                formatOnFocus,
                initialValue: value,
                isPending: false,
                parse,
                syncErrors: getSyncErrors(initialValue),
                touched: initialTouched,
                touchedAfterClick,
                value: initialValue,
                visited: initialVisited,
              },
              formName,
              name,
            })
          );
        });
      }
    }, [formName]);

    useEffect(() => {
      if (field) {
        updateAsyncValidators(value);
        updateSyncValidators(value);
      }
    }, [touched, value, visited]);

    if (children) {
      return <>{children(metaProps, wrapperRestProps)}</>;
    }

    if (Component) {
      return <Component meta={metaProps} {...wrapperRestProps} />;
    }

    return null;
  }
);

export default Field;
