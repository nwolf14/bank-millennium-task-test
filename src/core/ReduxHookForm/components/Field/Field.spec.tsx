import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import Checkbox from '../../../../shared/ComponentsUI/components/Checkbox/Checkbox';
import Field from './Field';
import TextField from '../../../../shared/ComponentsUI/components/TextField/TextField';

// others
import { InputValueType } from './enums';

const stateMock = {
  reduxHookForm: {
    testForm: {
      asyncTimeDelay: 0,
      error: '',
      fields: {
        testField: {
          active: false,
          asyncErrors: [],
          initialValue: '',
          isPending: false,
          parse: (value: any) => value,
          syncErrors: [],
          touched: false,
          touchedAfterClick: false,
          value: '',
          visited: false,
        },
      },
      isPending: false,
      isValid: false,
    },
  },
};

jest.mock('lodash', () => ({
  ...(jest.requireActual('lodash') as Object),
  defer: (callback: any) => callback(),
}));

describe('Field as Component', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <Field Component={TextField} formName="testForm" name="testField" />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Field as children', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <Field formName="unknown" name="unknown">
          {() => <TextField />}
        </Field>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Field as number type', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <Field
          Component={TextField}
          formName="testForm"
          inputValueType={InputValueType.number}
          name="testField"
          type="number"
        />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Field as boolean type', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <Field
          Component={Checkbox}
          formName="testForm"
          inputValueType={InputValueType.boolean}
          name="testField"
          type="number"
        />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Field as null', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <Field formName="unknown" name="unknown" />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
