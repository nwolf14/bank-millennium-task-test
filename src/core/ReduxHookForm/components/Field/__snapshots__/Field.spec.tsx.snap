// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field as Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TextField TextField--margin-bottom"
  >
    <div
      class="TextFieldInputWrapper TextFieldInputWrapper--medium"
    >
      <input
        class="TextFieldInput TextFieldInput--medium"
        data-testid="input"
        name="testField"
        placeholder=""
        style="height: 24px;"
        type="text"
        value=""
      />
      <label
        class="TextFieldLabel TextFieldLabel--medium"
      />
      <div
        class="TextFieldFilling"
      >
        <legend
          class="TextFieldFilling__legend"
        >
           
        </legend>
      </div>
    </div>
    <div
      class="ValidationMessage"
    />
  </div>
</DocumentFragment>
`;

exports[`Field as boolean type matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Checkbox Checkbox--margin-bottom"
  >
    <div
      class="Checkbox__wrapper"
    >
      <input
        class="CheckboxInput CheckboxInput--medium"
        data-testid="input"
        type="checkbox"
      />
      <div
        class="CheckboxIconWrapper CheckboxIconWrapper--medium"
      >
        <svg
          class="CheckboxIconWrapper__unchecked-icon"
        >
          uncheck.svg
        </svg>
      </div>
    </div>
    <div
      class="ValidationMessage"
    />
  </div>
</DocumentFragment>
`;

exports[`Field as children matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TextField TextField--margin-bottom"
  >
    <div
      class="TextFieldInputWrapper TextFieldInputWrapper--medium"
    >
      <input
        class="TextFieldInput TextFieldInput--medium"
        data-testid="input"
        name=""
        placeholder=""
        style="height: 24px;"
        type="text"
        value=""
      />
      <label
        class="TextFieldLabel TextFieldLabel--medium"
      />
      <div
        class="TextFieldFilling"
      >
        <legend
          class="TextFieldFilling__legend"
        >
           
        </legend>
      </div>
    </div>
    <div
      class="ValidationMessage"
    />
  </div>
</DocumentFragment>
`;

exports[`Field as null matches snapshot 1`] = `<DocumentFragment />`;

exports[`Field as number type matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TextField TextField--margin-bottom"
  >
    <div
      class="TextFieldInputWrapper TextFieldInputWrapper--medium"
    >
      <input
        class="TextFieldInput TextFieldInput--medium"
        data-testid="input"
        name="testField"
        placeholder=""
        style="height: 24px;"
        type="number"
        value=""
      />
      <label
        class="TextFieldLabel TextFieldLabel--placement-top TextFieldLabel--medium"
      />
      <div
        class="TextFieldFilling"
      >
        <legend
          class="TextFieldFilling__legend TextFieldFilling__legend--focus"
        >
           
        </legend>
      </div>
    </div>
    <div
      class="ValidationMessage"
    />
  </div>
</DocumentFragment>
`;
